### PR TITLE
GCE: Use network project id for firewall/route mgmt and zone listing

### LINKED ISSUE #48515: GCE: Project Id used for firewall mgmt instead of network project Id
In the GCE Cloudprovider package, firewall should be created in the project which contains the network.  Currently, the compute project Id is used. This is a broken feature of K8s 1.7.0.

### LINKED ISSUE #48521: GCE: Config's network name ignored when it's not a url
#47881 broke support for specifying network name.

### PR DESCRIPTION
**Pull request by nicksardo**
- Introduces a new environment variable for plumbing the network project id which will be used for firewall and route management. fixes #48515
- onXPN is determined by metadata if config is not specified
- Split `if` conditions: fixes #48521
- Remove `getNetworkNameViaAPICall` which was used as a last resort for the `networkURL` (if empty) which was previously filled with the metadata network project & name.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/cloudprovider/providers/gce/gce.go [hunk 1]
```diff
@@ -132,6 +134,7 @@ type Config struct {
 		TokenURL           string   `gcfg:"token-url"`
 		TokenBody          string   `gcfg:"token-body"`
 		ProjectID          string   `gcfg:"project-id"`
+		NetworkProjectID   string   `gcfg:"network-project-id"`
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce.go]: might want to have a comment here why this is different than ProjectID

File: pkg/cloudprovider/providers/gce/gce.go [hunk 2]
```diff
@@ -157,26 +160,38 @@ func (g *GCECloud) GetComputeService() *compute.Service {
 // newGCECloud creates a new instance of GCECloud.
 func newGCECloud(config io.Reader) (*GCECloud, error) {
 	apiEndpoint := ""
-	projectID, zone, err := getProjectAndZone()
+
+	// projectNumber is a project identifier known to be numeric
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce.go]: This comment is somewhat confusing. Probably should expand to something like:
```
// projectNumber is the numeric identifier. Note: there is also a unique string-based project identifier as well (see https://cloud.google.com/resource-manager/docs/creating-managing-projects#identifying_projects)
```

File: pkg/cloudprovider/providers/gce/gce.go [hunk 3]
```diff
@@ -132,6 +134,7 @@ type Config struct {
 		TokenURL           string   `gcfg:"token-url"`
 		TokenBody          string   `gcfg:"token-body"`
 		ProjectID          string   `gcfg:"project-id"`
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce.go]: Add comment `// ProjectID can either be the numeric or string-based unique identifier that starts with [a-z]`

File: pkg/cloudprovider/providers/gce/gce.go [hunk 4]
```diff
@@ -157,26 +160,38 @@ func (g *GCECloud) GetComputeService() *compute.Service {
...
+	// networkProjectNumber is a project identifier known to be numeric
+	networkProjectNumber, networkName, err := getNetworkProjectAndNameViaMetadata()
 	if err != nil {
 		return nil, err
 	}
-	networkURL := gceNetworkURL(apiEndpoint, projectID, networkName)
+	// Default networkProjectID to known network project number
+	networkProjectID := networkProjectNumber
+
+	// Rebuild the network URL
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce.go]: delete this comment?

File: pkg/cloudprovider/providers/gce/gce.go [hunk 5]
```diff
@@ -193,18 +208,36 @@ func newGCECloud(config io.Reader) (*GCECloud, error) {
 		if cfg.Global.ProjectID != "" {
 			projectID = cfg.Global.ProjectID
 		}
+		if cfg.Global.NetworkProjectID != "" {
+			networkProjectID = cfg.Global.NetworkProjectID
+		}
+
+		// Determine if cluster is on shared VPC network based off config
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce.go]: [minor] seems clearer:

... VPC network *according to config*

File: pkg/cloudprovider/providers/gce/gce.go [hunk 6]
```diff
@@ -46,6 +46,8 @@ import (
 const (
 	ProviderName = "gce"
 
+	defaultAPIEndpoint = "https://www.googleapis.com/compute/v1/"
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce.go]: [minor] looks a bit better below if leave off the trailing '/'

File: pkg/cloudprovider/providers/gce/gce_zones.go [hunk 7]
```diff
@@ -44,13 +44,13 @@ func (gce *GCECloud) GetZone() (cloudprovider.Zone, error) {
...
+	list, err := gce.service.Zones.List(gce.networkProjectID).Filter(filter).Do()
 	if err != nil {
 		return nil, mc.Observe(err)
 	}
 	return list.Items, mc.Observe(err)
 }
 
 func (gce *GCECloud) getRegionLink(region string) string {
-	return fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%v/regions/%v", gce.projectID, region)
+	return fmt.Sprintf("https://www.googleapis.com/compute/v1/projects/%v/regions/%v", gce.networkProjectID, region)
```
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce_zones.go]: fix API endpoint?

File: pkg/cloudprovider/providers/gce/gce.go (hunk 1 above)
Comment (bowei) [inline pkg/cloudprovider/providers/gce/gce.go]: Add documentation about which objects are associated with instances vs network...

**Review by bowei — COMMENTED** [review]
Mostly comments about comments

File: pkg/cloudprovider/providers/gce/gce_zones.go (hunk 7 above)
Comment (nicksardo) [inline pkg/cloudprovider/providers/gce/gce_zones.go]: leaving to #48568

**Comment by sakshamsharma** [comment]
#48574 uses `gce.GetProjectID()` which is to be removed by this PR. Please update that code as well, if it gets merged before this one. If this one gets merged first, I'll update the other one.

**Pull request merged**